### PR TITLE
[CORDA-1941]: Server-side draining node shutdown.

### PR DESCRIPTION
**Pull request by sollecitom**
https://r3-cev.atlassian.net/browse/CORDA-1941

### CODE REVIEW

**Comment by sollecitom** [comment]
@tlil , I ran the changes about `isSsh` in Shell with @szymonsztuka , and he agrees that they're good.

File: node/src/main/kotlin/net/corda/node/services/persistence/NodePropertiesPersistentStore.kt [hunk 1]
```diff
@@ -57,12 +57,14 @@ class FlowsDrainingModeOperationsImpl(readPhysicalNodeId: () -> String, private
 
     override val values = PublishSubject.create<Pair<Boolean, Boolean>>()!!
 
-    override fun setEnabled(enabled: Boolean) {
+    override fun setEnabled(enabled: Boolean, propagateChange: Boolean) {
         var oldValue: Boolean? = null
```
[the remark this answers is not in the document]
Comment (sollecitom) [reply node/src/main/kotlin/net/corda/node/services/persistence/NodePropertiesPersistentStore.kt]: Well, perhaps, but nothing to do with my changes haha. I'll do it anyway.

File: node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt [hunk 2]
```diff
@@ -49,7 +53,24 @@ internal class CordaRPCOpsImpl(
...
+        private val logger = loggerFor<CordaRPCOpsImpl>()
+    }
+
+    private val drainingShutdownHook = AtomicReference<Subscription?>()
+
+    init {
+        services.nodeProperties.flowsDrainingMode.values.filter { it.isDisabled() }.subscribe({
+            cancelDrainingShutdownHook()
+        }, {
+            // Nothing to do in case of errors here.
```
[the remark this answers is not in the document]
Comment (sollecitom) [reply node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt]: Here we really don't care, if there's an error, it's handled at the appropriate level, but this consumer doesn't care.

File: node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt [hunk 3]
```diff
@@ -271,18 +292,44 @@ internal class CordaRPCOpsImpl(
...
+
+    override fun shutdown() = terminate(false)
+
+    override fun terminate(drainPendingFlows: Boolean) {
+
+        if (drainPendingFlows) {
+            logger.info("Waiting for pending flows to complete before shutting down.")
+            setFlowsDrainingModeEnabled(true)
+            drainingShutdownHook.set(pendingFlowsCount().updates.doOnCompleted { setPersistentDrainingModeProperty(false, false) }.doOnCompleted(::cancelDrainingShutdownHook).doOnCompleted { logger.info("No more pending flows to drain. Shutting down.") }.doOnCompleted(shutdownNode::invoke).subscribe({
+                // Nothing to do on each update here, only completion matters.
```
[the remark this answers is not in the document]
Comment (sollecitom) [reply node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt]: I think `pendingFlowsCount()` does that already, I'll check.

File: tools/shell/src/main/kotlin/net/corda/tools/shell/InteractiveShell.kt [hunk 4]
```diff
@@ -467,50 +467,56 @@ object InteractiveShell {
...
-                            System.out.println("Shutting down the node via remote SSH session (it may take a while)")
+            cordaRPCOps.terminate(true)
+
+            val latch = CountDownLatch(1)
+            cordaRPCOps.pendingFlowsCount().updates.doOnError { error ->
+                log.error(error.message)
+                throw error
+            }.doAfterTerminate(latch::countDown).subscribe(
+                    // For each update.
+                    { (first, second) -> display { println("...remaining: $first / $second") } },
```
[the remark this answers is not in the document]
Comment (sollecitom) [reply tools/shell/src/main/kotlin/net/corda/tools/shell/InteractiveShell.kt]: `println` here is a different "println" from `System.out.println`, it's a function exposed by `RenderPrintWriter`, which is the way Shell decides what to do with messages to log.

Also, I haven't changed that, merely adapted the existing logic to work within the `Observable`'s subscriber.

[the remark this answers is not in the document]
Comment (tlil) [reply tools/shell/src/main/kotlin/net/corda/tools/shell/InteractiveShell.kt]: OK, might be worth making sure the error gets logged though (if it isn't already). :-)

File: node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (sollecitom) [reply node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt]: I was mistaken. Added the logging at call site in `CordaRPCOpsImpl`, in a similar way `InteractiveShell` does it./

**Pull request merged**